![image](competi.png)

#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

In [1]:
import numpy as np
import pandas as pd

import urllib.request
from PIL import Image

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import LabelEncoder
from sklearn.svm import SVR

In [2]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
df = pd.read_csv("diamonds_train.csv", index_col=0)

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40345 entries, 0 to 40344
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   carat    40345 non-null  float64
 1   cut      40345 non-null  object 
 2   color    40345 non-null  object 
 3   clarity  40345 non-null  object 
 4   depth    40345 non-null  float64
 5   table    40345 non-null  float64
 6   x        40345 non-null  float64
 7   y        40345 non-null  float64
 8   z        40345 non-null  float64
 9   price    40345 non-null  int64  
dtypes: float64(6), int64(1), object(3)
memory usage: 3.4+ MB


In [4]:
le = LabelEncoder()

In [5]:
df["clarity"] = le.fit_transform(df["clarity"])
df["color"] = le.fit_transform(df["color"])
df["cut"] = le.fit_transform(df["cut"])

In [6]:
df.corr()

carat       cut     color   clarity     depth     table         x  \
carat    1.000000  0.018551  0.285737 -0.213653  0.029414  0.181273  0.975077   
cut      0.018551  1.000000 -0.000943  0.026981 -0.195392  0.149833  0.023399   
color    0.285737 -0.000943  1.000000 -0.030010  0.048637  0.022557  0.263915   
clarity -0.213653  0.026981 -0.030010  1.000000 -0.054952 -0.088440 -0.225114   
depth    0.029414 -0.195392  0.048637 -0.054952  1.000000 -0.288987 -0.023389   
table    0.181273  0.149833  0.022557 -0.088440 -0.288987  1.000000  0.193986   
x        0.975077  0.023399  0.263915 -0.225114 -0.023389  0.193986  1.000000   
y        0.944445  0.028351  0.255262 -0.215307 -0.027880  0.180449  0.966827   
z        0.967289  0.002183  0.266589 -0.226675  0.097399  0.153175  0.985080   
price    0.921887  0.039499  0.166340 -0.070364 -0.009228  0.127338  0.883163   

                y         z     price  
carat    0.944445  0.967289  0.921887  
cut      0.028351  0.002183  0.039499  
color    0.255262  0.266589  0.166340  
clarity -0.215307 -0.226675 -0.070364  
depth   -0.027880  0.097399 -0.009228  
table    0.180449  0.153175  0.127338  
x        0.966827  0.985080  0.883163  
y        1.000000  0.959957  0.857564  
z        0.959957  1.000000  0.872855  
price    0.857564  0.872855  1.000000

In [7]:
df.drop("depth",inplace=True,axis=1)
df.drop("table",inplace=True,axis=1)


### 1. Definir X e y

In [8]:
df.columns

Index(['carat', 'cut', 'color', 'clarity', 'x', 'y', 'z', 'price'], dtype='object')

In [9]:

X = np.array(df[["carat","cut","color","clarity","x","y","z"]])
y = np.array(df["price"])
X.shape

(40345, 7)

In [10]:
y.shape

(40345,)

### 2. Dividir X_train, X_test, y_train, y_test

In [11]:
# el test_size 0.20 y el random_state puede cambiar
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 7)

In [12]:
X_train

array([[1.09, 3.  , 1.  , ..., 6.75, 6.61, 4.07],
       [1.5 , 3.  , 4.  , ..., 7.27, 7.34, 4.58],
       [0.41, 2.  , 0.  , ..., 4.77, 4.8 , 2.97],
       ...,
       [0.34, 3.  , 2.  , ..., 4.59, 4.54, 2.72],
       [0.5 , 0.  , 0.  , ..., 4.95, 4.84, 3.31],
       [0.91, 1.  , 3.  , ..., 6.06, 6.1 , 3.9 ]])

### 3. Asignar el modelo (vacío) a una variable
Aquí meteríais los parámetros. 

**Consejo**: Usa GridSearch y vuélvete loca o loco probando modelos e hiperparámetros.

In [13]:
svr_rbf = SVR(kernel='poly', C=100, gamma=0.1, epsilon=.1)
model = svr_rbf.fit(X, y)
print("model.score:", model.score(X, y))
y_prediction = model.predict(X)

model.score: 0.9316670717262875


In [22]:
from sklearn.metrics import r2_score, mean_absolute_error, accuracy_score

In [24]:
r2_score(y_train, y_prediction)

ValueError: Found input variables with inconsistent numbers of samples: [32276, 40345]

### 4. Entrenar el modelo con X_train, y_train

### 5. Predecir con el modelo ya entrenado con X_test

In [14]:
predictions = y_prediction
print(predictions)

[7050.79143411  691.81463135 1090.44205898 ... 1566.34751069 5538.6974816
 1448.99263077]


### 6. Sacar métricas, valorar el modelo 

Recuerdo que en la competición se va a evaluar con la métrica de RMSE.

In [54]:
np.sqrt(mean_squared_error(y_train, predictions))

ValueError: Found input variables with inconsistent numbers of samples: [32276, 40345]

-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

_Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `diamonds_test.csv` nos arriesgamos a que digamos, no salga lo esperado._

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `diamonds_train.csv` al completo.

**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**

Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `diamonds_test.csv` para predecir.

**¿De dónde saco `diamonds_test.csv`?**

In [26]:
X_pred = pd.read_csv("diamonds_test.csv", index_col = 0)
X_pred.head()

carat        cut color clarity  depth  table     x     y     z
0   0.30      Ideal     H     SI2   60.0   56.0  4.41  4.43  2.65
1   0.34      Ideal     D      IF   62.1   57.0  4.52  4.46  2.79
2   1.57  Very Good     I     VS2   60.3   58.0  7.58  7.55  4.56
3   0.31      Ideal     H     VS2   61.8   57.0  4.32  4.36  2.68
4   1.51       Good     I    VVS1   64.0   60.0  7.26  7.21  4.63

In [27]:
X_pred.drop("depth",inplace=True,axis=1)
X_pred.drop("table",inplace=True,axis=1)


In [28]:
X_pred["clarity"] = le.fit_transform(X_pred["clarity"])
X_pred["color"] = le.fit_transform(X_pred["color"])
X_pred["cut"] = le.fit_transform(X_pred["cut"])

In [29]:
X_pred = np.array(X_pred[["carat","cut","color","clarity","x","y","z"]])

In [30]:
X.shape

(40345, 7)

In [31]:
X_pred.shape

(13449, 7)

**¿Por qué da error?** 

In [32]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([  442.33516771,  1267.60470419, 10368.79553317, ...,
       12961.88585305, 13259.73986808,   812.05749573])

## IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 13449 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `diamonds_test.csv`?**

In [33]:
X_pred = np.array(X_pred[["carat","cut","color","clarity","x","y","z"]])

C:\Users\34606\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:1: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  """Entry point for launching an IPython kernel.


IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

Porque:
    - SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 3 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS

In [ ]:
X_pred

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [ ]:
# respuestas

In [ ]:
predictions_submit = model.predict(X_pred)
predictions_submit

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

**¿De dónde saco `sample_submission.csv`?**

In [ ]:
sample = pd.read_csv("sample_submission.csv")

In [ ]:
sample.head()

In [ ]:
sample.shape

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [48]:
submission = pd.DataFrame({"id": range(len(predictions_submit)), "price": predictions_submit})

In [49]:
submission.head()

id         price
0   0    442.335168
1   1   1267.604704
2   2  10368.795533
3   3    574.889976
4   4   9209.587248

In [50]:
submission.shape

(13449, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [51]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.id.all() == sample.id.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [52]:
help(chequeator)

Help on function chequeator in module __main__:

chequeator(df_to_submit)
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.



In [53]:
chequeator(submission)

You're ready to submit!
